services/repository/files/update.go

```diff
@@ -232,176 +295,215 @@ func CreateOrUpdateRepoFile(ctx context.Context, repo *repo_model.Repository, do
 
 		}
 
-		if !opts.IsNewFile {
-			fromEntry, err := commit.GetTreeEntryByPath(fromTreePath)
-			if err != nil {
-				return nil, err
-			}
-			if opts.SHA != "" {
-				// If a SHA was given and the SHA given doesn't match the SHA of the fromTreePath, throw error
-				if opts.SHA != fromEntry.ID.String() {
-					return nil, models.ErrSHADoesNotMatch{
-						Path:       treePath,
-						GivenSHA:   opts.SHA,
-						CurrentSHA: fromEntry.ID.String(),
+		for _, file := range opts.Files {
+			fileOptions := filesOptions[file.TreePath]
```

Comment:
It seems that the `opts.Files[].TreePath` should also be applied to `CleanUploadFileName` ahead, otherwise the key might not match.